src/main/java/org/testng/DependencyMap.java

```diff
@@ -63,16 +61,32 @@ public ITestNGMethod getMethodDependingOn(String methodName, ITestNGMethod fromM
       // If they are in the same class hierarchy, they must belong to the same instance,
       // otherwise, it's a method depending on a method in a different class so we
       // don't bother checking the instance
-      if (fromMethod.getRealClass().isAssignableFrom(m.getRealClass())) {
-        if (m.getInstance() == fromMethod.getInstance()) return m;
-      } else {
+      if (isSameInstance(fromMethod, m) || belongToDifferentClassHierarchy(fromMethod, m) || hasNoInstance(fromMethod, m)) {
         return m;
       }
     }
     throw new TestNGException("Method \"" + fromMethod
         + "\" depends on nonexistent method \"" + methodName + "\"");
   }
 
+  private static boolean belongToDifferentClassHierarchy(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    return ! baseClassMethod.getRealClass().isAssignableFrom(derivedClassMethod.getRealClass());
+  }
+
+  private static boolean hasNoInstance(ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    Object baseInstance = baseClassMethod.getInstance();
+    Object derivedInstance = derivedClassMethod.getInstance();
+    return derivedInstance != null || baseInstance != null;
```

Comment:
`return derivedInstance == null && baseInstance == null;` looks easier to understand according to the method name.

Reply:
@juherr - The logic cannot be changed, else it will cause test failures (especially the new one that was added and what this PR intends to fix). The method name actually meant that either `baseClassMethod` (or) `derivedClassMethod` *has no instance*

Reply:
Ok, but `derivedInstance != null || baseInstance != null` means one of the 2 has an instance, right?

Reply:
Yes, that's one interpretation. The other would be, that one of them has no instance when `!=null` fails. Like I said, I think the method name is a bit confusing. Am open to any other suggestions

Reply:
In fact, the name should be `hasInstance` because true is returned if `baseClassMethod` or `derivedClassMethod` have an instance.

Reply:
@juherr - I have renamed the method to `hasInstance`